rholang/src/main/scala/coop/rchain/rholang/interpreter/accounting/package.scala

```diff
@@ -1,23 +1,56 @@
 package coop.rchain.rholang.interpreter
 
-import cats.Monad
+import cats._
+import cats.data._
+import cats.effect.Sync
+import cats.effect.concurrent.Ref
 import cats.implicits._
-import cats.mtl.MonadState
+import cats.mtl._
 
 import coop.rchain.rholang.interpreter.errors.OutOfPhlogistonsError
 
 package object accounting extends Costs {
 
-  type _cost[F[_]] = MonadState[F, Cost]
-  def _cost[F[_]](implicit ev: _cost[F]): _cost[F] = ev
+  type _cost[F[_]]       = MonadState[F, Cost]
+  type _costWriter[F[_]] = FunctorTell[F, Chain[Cost]]
+
+  def _cost[F[_]](implicit ev: _cost[F]): _cost[F]                   = ev
+  def _costWriter[F[_]](implicit ev: _costWriter[F]): _costWriter[F] = ev
 
   def charge[F[_]: Monad](
       amount: Cost
-  )(implicit cost: _cost[F], error: _error[F]): F[Unit] =
+  )(implicit cost: _cost[F], error: _error[F], writer: _costWriter[F]): F[Unit] =
     for {
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
+      _ <- writer.tell(Chain.one(amount))
       _ <- cost.modify(_ - amount)
       _ <- error.ensure(cost.get)(OutOfPhlogistonsError)(_.value >= 0)
     } yield ()
 
+  def costWriter[M[_]: Sync](): FunctorListen[M, Chain[Cost]] =
+    new DefaultFunctorListen[M, Chain[Cost]] {
+      override val functor: Functor[M]  = implicitly[Functor[M]]
+      private val ref                   = Ref.unsafe(Chain.empty[Cost])
+      def tell(l: Chain[Cost]): M[Unit] = ref.modify(c => (c.concat(l), ()))
+      def listen[A](fa: M[A]): M[(A, Chain[Cost])] =
+        for {
+          a <- fa
+          r <- ref.get
+        } yield ((a, r))
+    }
+
+  def noOpCostWriter[M[_]: Applicative]: _costWriter[M] =
+    new DefaultFunctorTell[M, Chain[Cost]] {
+      override val functor: Functor[M]  = implicitly[Functor[M]]
+      def tell(l: Chain[Cost]): M[Unit] = Applicative[M].pure(())
+    }
+
+  implicit def ntCostWriter[F[_]: Monad: _costWriter, G[_]: Functor](
+      nt: F ~> G
+  ): _costWriter[G] =
```

Comment:
Why is this needed? I think it might no logner be needed after we make `_cost` do the writing transparently.